server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -197,6 +197,8 @@ private synchronized void syncSegments(boolean isRetry) {
             logger.info("syncSegments is only supported for InternalEngine, called with {}. Skipping", indexShard.getEngine());
             return;
         }
+        ReplicationCheckpoint checkpoint = indexShard.getLatestReplicationCheckpoint();
+        indexShard.onCheckpointPublished(checkpoint);
```

Comment:
Do we need for the upload to complete before invoking `onCheckpointPublished`?

Reply:
Both these operations should ideally occur together otherwise looks like there is something wrong if we are invoking it independently.

https://github.com/opensearch-project/OpenSearch/blob/230a06daee9c8484e38d6aad867a98c3c10c2485/server/src/main/java/org/opensearch/indices/replication/checkpoint/SegmentReplicationCheckpointPublisher.java#L35-L38

Reply:
> Taking a stab on answering this - IIRC onCheckpointPublished is invoked before segments upload so that the replication lag can be determined correctly and is wholesome lag from the point of refresh happening on primary and the segments getting copied over on replicas. @ankitkala Do add anything else if missed.
> 
> @Bukhtawar this is a backport PR. This is blocking couple of auto backports of other PRs when backport label is added.

@Bukhtawar `onCheckpointPublished` acts as a signal to replicas for syncing from remote. So, we do need to wait for the upload to finish.

Reply:
@ankitkala But in `RemoteStoreRefreshListener` onCheckpointPublished is called before the checkpoint is published. The flow is, onCheckpointPublished -> upload segments -> publish checkpoint

Reply:
Tracking issue: https://github.com/opensearch-project/OpenSearch/issues/8750